### DISCUSSION TITLE
DateInput value as string

### DISCUSSION THREAD

**Discussion by joshxyzhimself** [mantinedev/mantine#6441] [Q&A] [answered]
Hello to my mantine overlords,

![image](https://github.com/mantinedev/mantine/assets/60829711/fdee5094-a193-4c28-a3b1-967ed22ecadb)

It appears that `valueFormat` only affects the value displayed to the user,
not the value output by the element itself which is `Date` instead of `string`.

Context:

- In database we accept string in `iso 8601 date` format.
- In forms we use zod validation of `z.string().date().nullable()` which is straightforward on the form-level and the rest-api level since in json we can only encode strings not date objects.
- DateInput component produces `Date` object output, and there is no option (that im aware of) to produce just iso date strings.
- Sure one can add a manual step of capture the Date object and manually transform it to string, but we are using mantine forms + zod which makes me confused how to properly approach it haha, any ideas?

Thank you thank you!

---

I ended up settling with:

```
<TextInput type="date" label="Name" size="xs" required />
```

**Comment by rtivital** [comment, marked as answer]
In 7.x: `@mantine/dates` package works with Date objects, components do not support strings as values
In 8.x and higher versions: All dates packages were migrated to use strings instead of dates. You need to use date in `YYYY-MM-DD` or `YYYY-MM-DD HH:mm:ss` depending on whether the time is needed

**Comment by uncomplexity** [comment]
here's what worked for us

bit cumbersome but worth it, as DateInput looks better than TextInput

```ts
// turns string | null to Date | null
type FieldsToDateNull<T, K extends keyof T> = {
  [P in keyof T]: P extends K ? Date | null : T[P];
};

// from server
type User = {
  // other values
  birthDate: string | null
};

// it's now Date | null
type UserPersonal = FieldsToDateNull<User, "birthDate">

// can now use it like this
const personalForm = useForm<UserPersonal>({
  mode: 'uncontrolled',
  initialValues: {
    // other values
    birthDate: null,
  },
});


// when loading data, convert it to Date | null
const personalInitialValues = pick(data, [
  // other values
  'birthDate',
]);
personalFormSetInitialValues({
  ...personalInitialValues, // spread the other values
  birthDate: personalInitialValues.birthDate
    ? DateTime.fromISO(
        personalInitialValues.birthDate,
      ).toJSDate()
    : null,
});
personalFormReset();


// when saving data, convert it to string | null
const data = {
    ...personalValues, // spread the other values
    birthDate: personalValues.birthDate
      ? DateTime.fromJSDate(
          personalValues.birthDate,
        ).toISODate()
      : null,
  },
};
```
